### PR TITLE
Add opentelemetry metrics integration on the same prometheus endpoint [PLEN-4]

### PR DESCRIPTION
**Pull request by nicu-da**
- The opentelemetry metrics are exposed only as part of the prometheus reporter.
- The dropwizard metrics and opentelemetry metrics work together
- Further integration with the metrics API will be done once https://github.com/digital-asset/daml/pull/15231 is merged

changelog_begin
changelog_end

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/metrics/MetricsOwner.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        new MetricRegistry
+    }
+    val reporter = Option.when(config.enabled) {
+      val runningReporter = config.reporter
+        .register(meterRegistry)
+      runningReporter.start(config.reportingInterval.toMillis, TimeUnit.MILLISECONDS)
+      runningReporter
+    }
+
+    meterRegistry.registerAll(new JvmMetricSet)
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/metrics/MetricsOwner.scala:35]: Need to think about how we will get JVM metrics into the opentelemetry to allow a like-for-like experience.

Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/metrics/MetricsOwner.scala:35]: Additionally and more importantly by moving this line from sandbox-on-x to this shared class canton will pick up on this.  This is not desirable as canton have a config flag that determines if JVM metrics should be added.  So probably best to add  a helper method that can be:
- built out with support for opentelemtry
- called from sandbox and canton as required.

**Review by simonmaxen-da — COMMENTED** [review]
Comments added.

File: ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala [hunk 2]
```diff
@@ -56,7 +57,7 @@ object IndexMetadata {
       executionContext: ExecutionContext,
       loggingContext: LoggingContext,
   ) = {
-    val metrics = new Metrics(new MetricRegistry)
+    val metrics = new Metrics(new MetricRegistry, GlobalOpenTelemetry.getMeter("daml"))
```
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala:60]: Sure, I can extract this into a helper.

File: ledger/participant-integration-api/src/main/scala/platform/metrics/MetricsOwner.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+package com.daml.platform.metrics
+
+import java.util.concurrent.TimeUnit
+
+import com.codahale.metrics.{MetricRegistry, SharedMetricRegistries}
+import com.daml.ledger.resources.{Resource, ResourceContext, ResourceOwner}
+import com.daml.metrics.{JvmMetricSet, Metrics}
+import com.daml.platform.config.MetricsConfig
+import com.daml.platform.config.MetricsConfig.MetricRegistryType
```
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/metrics/MetricsOwner.scala:12]: Not sure I follow. 
Couldn't find a the scaladoc for it in `MetricsConfig` file. 
The code is just moved from the sandbox on x runner so no actual changes are done to the way we build the registry.

File: ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala:60]: Actually this is the only "daml" meter that I see. All the others are for "test"

File: ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import com.daml.ledger.resources.{Resource, ResourceContext, ResourceOwner}
+import com.daml.metrics.MetricsReporter.Prometheus
+import io.opentelemetry.api.metrics.Meter
+import io.opentelemetry.exporter.prometheus.PrometheusCollector
+import io.opentelemetry.sdk.metrics.SdkMeterProvider
+
+import scala.annotation.nowarn
+import scala.concurrent.Future
+
+@nowarn("msg=deprecated")
```
Comment (simonmaxen-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala:15]: Could you add some text explaining why we are introducing a deprecated class.  Note that [PrometheusCollector](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/prometheus/src/integrationTest/java/io/opentelemetry/exporter/prometheus/PrometheusCollector.java) exists in `integrationTest` and is also deprecated.

Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala:15]: Good point, will add to the code, it was initially documented in the ticket itself.

Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala:15]: Added a comment about the usage of the deprecated class

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/MetricsOwner.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import io.opentelemetry.api.metrics.Meter
+
+import scala.concurrent.Future
+
+case class MetricsOwner(meter: Meter, config: MetricsConfig, name: String)
+    extends ResourceOwner[Metrics] {
+  override def acquire()(implicit
+      context: ResourceContext
+  ): Resource[Metrics] = {
+    val meterRegistry = config.registryType match {
```
Comment (simonmaxen-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/MetricsOwner.scala:22]: Should be metricRegistry ?

File: ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    val meterProvider = if (enabled && reporter.exists(_.isInstanceOf[Prometheus])) {
+      meterProviderBuilder.registerMetricReader(PrometheusCollector.create()).build()
+    } else meterProviderBuilder.build()
+    Resource(
+      Future(
+        meterProvider.meterBuilder("daml").build()
+      )
+    ) { _ =>
+      Future {
+        meterProvider.close()
```
Comment (simonmaxen-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/OpenTelemetryMeterOwner.scala:37]: Based on our usage of `CollectorRegistry.defaultRegistry.unregister(exports)` in `PrometheusReporter` to follow that pattern `PrometheusCollector.shutdown` should be called.

**Review by simonmaxen-da — APPROVED** [review]
This follows the pattern used in the `PrometheusReporter` / `DropwizardExports` of using the registering collectors with a global `CollectorRegistry.defaultRegistry` that is also used by the `io.prometheus.client.expoerter.HttpServer` by default.

I would prefer to see us switching to the OpenTelemetry `PrometheusHttpServer` which is constructed with a `MetricsProducer`.  Either way this PR is a step in the right direction (towards opentelemetry).

**Comment by nicu-da** [comment]
> I would prefer to see us switching to the OpenTelemetry `PrometheusHttpServer` which is constructed with a `MetricsProducer`. Either way this PR is a step in the right direction (towards opentelemetry).

Yes, unfortunately in the future we will have to find a different way to expose the metrics. 
We will revisit the way we integrate these libraries in the future.

**Pull request merged**